### PR TITLE
chore: implements top-level-await

### PR DESCRIPTION
**Pull request by KFlash**

### CODE REVIEW

**Comment by KFlash** [comment]
@aladdin-add  Added bunch of TC39 tests. In script mode there are no changes.  

We are using the `TopLevel` bitwise mask, and when entering function scope this mask is unset unless parsing as an async func.

I'm not sure how this should work on nested functions?

**Comment by KFlash** [comment]
@aladdin-add This is ready now, so if you accept it you can merge it.

File: test/__snapshot__/parser/module/top-level-await/export-var-await-expr-identifier.md [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+### Printed
+
+```javascript
+
+var foo = 1;
+ var name1 = await foo;
+ var {
+  x = await foo
+} = {};
+```
```
Comment (aladdin-add) [inline test/__snapshot__/parser/module/top-level-await/export-var-await-expr-identifier.md:273]: the printed code is incorrect.

File: test/__snapshot__/parser/module/top-level-await/for-await-module.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    "start": 0,
+    "end": 25
+}
+```
+
+### Printed
+
+```javascript
+
+for await ( of b);
```
Comment (aladdin-add) [inline test/__snapshot__/parser/module/top-level-await/for-await-module.md:101]: the printed code is incorrect.

File: test/__snapshot__/parser/module/top-level-await/inside-arrow.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+### Printed
+
+```javascript
+
+```
+
+### Diagnostics
+
+```javascript
+✖ The parser expected to find a '}' to match the '{' token here - start: 14, end: 14
```
Comment (aladdin-add) [inline test/__snapshot__/parser/module/top-level-await/inside-arrow.md:106]: nits: there is a room to improve the message, e.g. "expect 'await' in an async function."

**Review by aladdin-add — CHANGES_REQUESTED** [review]

File: test/__snapshot__/parser/module/top-level-await/inside-arrow.md (hunk 3 above)
Comment (KFlash) [reply test/__snapshot__/parser/module/top-level-await/inside-arrow.md]: I agree. It got caught in error recovery mode, and made some other tests fail. I have fixed it now.

File: test/__snapshot__/parser/module/top-level-await/export-var-await-expr-identifier.md (hunk 1 above)
Comment (KFlash) [reply test/__snapshot__/parser/module/top-level-await/export-var-await-expr-identifier.md:273]: what is missing here?

Comment (aladdin-add) [reply test/__snapshot__/parser/module/top-level-await/export-var-await-expr-identifier.md:273]: technically it's a valid code, but the input is:
```js
var foo = 1;
export var name1 = await foo;
export var { x = await foo } = {};
```

but the `export` is missing in the output.

Comment (KFlash) [reply test/__snapshot__/parser/module/top-level-await/export-var-await-expr-identifier.md:273]: Printing issue again?

File: test/__snapshot__/parser/module/top-level-await/for-await-module.md (hunk 2 above)
Comment (KFlash) [reply test/__snapshot__/parser/module/top-level-await/for-await-module.md]: It has been solved. It was all about wrong bitwise masks :(

**Comment by KFlash** [comment]
@aladdin-add Can you review again?

**Comment by KFlash** [comment]
@aladdin-add Can you merge this if you are okay with the changes?

**Review by aladdin-add — APPROVED** [review]
LGTM! 💯

**Pull request merged**